File: 2022/day3/day3.py
file_path = './input.txt'
with open(file_path, 'r') as f:
    lines = f.readlines()
    f.close()

p1 = 0
for line in lines:
    line = line.strip('\n')
    mid = int(len(line) / 2)
    uni = set(line[:mid]) & set(line[mid:])

    n = list(uni)[0]
    if n.islower():
        p1 += ord(n) - ord('a') + 1
    else:
        p1 += ord(n) - ord('A') + 27

print(f'[Part 1] : {p1}')

p2 = 0
for i in range(0, len(lines), 3):
    uni = set(lines[i].strip('\n')) & set(lines[i+1].strip('\n')) & set(lines[i+2].strip('\n'))
    n = list(uni)[0]
    if n.islower():
        p2 += ord(n) - ord('a') + 1
    else:
        p2 += ord(n) - ord('A') + 27

print(f'[Part 2] : {p2}')

File: 2024/Day03 Mull It Over/day3.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func readFile(fileName string) []string {
	data, err := os.ReadFile(fileName)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return []string{}
	}

	lines := strings.Split(string(data), "\n")
	return lines
}

func scan(s string) int {
	re := regexp.MustCompile(`mul\(\d+,\d+\)`)
	matches := re.FindAllString(s, -1)

	ans := 0
	for _, m := range matches {
		re = regexp.MustCompile(`mul\((\d+),(\d+)\)`)
		lst := re.FindStringSubmatch(m)
		a, _ := strconv.Atoi(lst[1])
		b, _ := strconv.Atoi(lst[2])

		ans += a * b
	}
	return ans
}

func scan2(s string) int {
	re := regexp.MustCompile(`mul\(\d+,\d+\)|don't\(\)|do\(\)`)
	matches := re.FindAllString(s, -1)
	//fmt.Println(matches)
	ans := 0
	isDisabled := false
	for _, m := range matches {
		if m[0:3] == "mul" {
			if isDisabled {
				continue
			}
			re = regexp.MustCompile(`mul\((\d+),(\d+)\)`)
			lst := re.FindStringSubmatch(m)
			a, _ := strconv.Atoi(lst[1])
			b, _ := strconv.Atoi(lst[2])
			ans += a * b
		} else if m[0:3] == "don" {
			isDisabled = true
		} else {
			isDisabled = false
		}
	}
	return ans
}

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	lines := readFile(filepath.Join(dir, "input.txt"))
	p1, p2 := 0, 0
	s := ""
	for _, line := range lines {
		s += line
	}
	p1 = scan(s)
	p2 = scan2(s)
	fmt.Printf("[Part 1]: %d\n", p1)
	fmt.Printf("[Part 2]: %d\n", p2)
}
